File: 085_BoundaryTraversal.py
"""
Problem: Boundary Traversal
Description: Print boundary nodes of a binary tree in anti-clockwise direction.

Approach: Print left boundary, leaves, then right boundary (bottom-up).
"""

from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isLeaf(node: TreeNode) -> bool:
    return not node.left and not node.right

def addLeftBoundary(root: TreeNode, result: List[int]):
    curr = root.left
    while curr:
        if not isLeaf(curr):
            result.append(curr.val)
        if curr.left:
            curr = curr.left
        else:
            curr = curr.right

def addLeaves(root: Optional[TreeNode], result: List[int]):
    if not root:
        return
    if isLeaf(root):
        result.append(root.val)
        return
    addLeaves(root.left, result)
    addLeaves(root.right, result)

def addRightBoundary(root: TreeNode, result: List[int]):
    curr = root.right
    temp = []
    while curr:
        if not isLeaf(curr):
            temp.append(curr.val)
        if curr.right:
            curr = curr.right
        else:
            curr = curr.left
    result.extend(reversed(temp))

def boundaryTraversal(root: Optional[TreeNode]) -> List[int]:
    result = []
    if not root:
        return result
    
    if not isLeaf(root):
        result.append(root.val)
    addLeftBoundary(root, result)
    addLeaves(root, result)
    addRightBoundary(root, result)
    
    return result

def main():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    
    print(boundaryTraversal(root))

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(h)
"""

File: 085_BoundaryTraversal.cpp
/*
Problem: Boundary Traversal
Description: Print boundary nodes of a binary tree in anti-clockwise direction.

Approach: Print left boundary, leaves, then right boundary (bottom-up).
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

bool isLeaf(TreeNode* node) {
    return !node->left && !node->right;
}

void addLeftBoundary(TreeNode* root, vector<int>& result) {
    TreeNode* curr = root->left;
    while (curr) {
        if (!isLeaf(curr)) result.push_back(curr->val);
        if (curr->left) curr = curr->left;
        else curr = curr->right;
    }
}

void addLeaves(TreeNode* root, vector<int>& result) {
    if (!root) return;
    if (isLeaf(root)) {
        result.push_back(root->val);
        return;
    }
    addLeaves(root->left, result);
    addLeaves(root->right, result);
}

void addRightBoundary(TreeNode* root, vector<int>& result) {
    TreeNode* curr = root->right;
    vector<int> temp;
    while (curr) {
        if (!isLeaf(curr)) temp.push_back(curr->val);
        if (curr->right) curr = curr->right;
        else curr = curr->left;
    }
    for (int i = temp.size() - 1; i >= 0; i--) {
        result.push_back(temp[i]);
    }
}

vector<int> boundaryTraversal(TreeNode* root) {
    vector<int> result;
    if (!root) return result;
    
    if (!isLeaf(root)) result.push_back(root->val);
    addLeftBoundary(root, result);
    addLeaves(root, result);
    addRightBoundary(root, result);
    
    return result;
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    
    vector<int> result = boundaryTraversal(root);
    for (int x : result) cout << x << " ";
    cout << endl;
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(h)
*/
